pkg/common/log/reopen.go

```diff
@@ -0,0 +1,110 @@
+package log
+
+import (
+	"context"
+	"io"
+	"os"
+	"os/signal"
+	"sync"
+	"syscall"
+)
+
+const (
+	_fileFlags    = os.O_APPEND | os.O_CREATE | os.O_WRONLY
+	_fileMode     = 0640
+	_reopenSignal = syscall.SIGUSR2
+)
+
+var _ ReopenableWriteCloser = (*ReopenableFile)(nil)
+
+// Reopener inspired by https://github.com/client9/reopen
+type Reopener interface {
+	Reopen() error
+}
+
+type ReopenableWriteCloser interface {
+	Reopener
+	io.WriteCloser
+}
+
+type ReopenableFile struct {
+	name string
+	f    *os.File
+	mu   sync.Mutex
+}
+
+func NewReopenableFile(name string) (*ReopenableFile, error) {
+	file, err := os.OpenFile(name, _fileFlags, _fileMode)
+	if err != nil {
+		return nil, err
+	}
+	return &ReopenableFile{
+		name: name,
+		f:    file,
+	}, nil
+}
+
+func (r *ReopenableFile) Reopen() error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	if r.f != nil {
+		if err := r.f.Close(); err != nil {
+			return err
+		}
+		r.f = nil
+	}
+	newFile, err := os.OpenFile(r.name, _fileFlags, _fileMode)
+	if err != nil {
+		r.f = nil
+		return err
+	}
+	r.f = newFile
+	return nil
+}
+
+func (r *ReopenableFile) Write(b []byte) (n int, err error) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	return r.f.Write(b)
+}
+
+func (r *ReopenableFile) Close() error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	return r.f.Close()
+}
+
+// Name implements part of os.FileInfo without needing a lock on the
+// underlying file.
+func (r *ReopenableFile) Name() string {
+	return r.name
+}
+
+// ReopenOnSignal returns a function compatible with RunTasks.
+func ReopenOnSignal(rwc ReopenableWriteCloser) func(context.Context) error {
```

Comment:
Yes, technically. That might be cleaner in test, but I think it obscures the intent.

Reply:
I'm not sure I follow. This is a function that reopens something that can be reopened on receipt of a signal. It doesn't care about writing and closing. Making the change also has the added benefit of removing functions from test fakes that aren't ever called. Otherwise, I don't see a point to the `Reopener` interface if we aren't going to use it anywhere.

Reply:
Thanks for your additional clarification. Yeah, I think I was over-indexed on aligning with the production use where this takes a `*ReopenableFile` (which happens to be a `ReopenableWriteCloser`). This is exactly the site which justifies the `Reopener` as a standalone interface (which I created instinctually). To require a `ReopenableWriteCloser` here means I only got that design 1/2 correct. ❤️